pygmt/helpers/decorators.py

```diff
@@ -469,3 +470,65 @@ def remove_bools(kwargs):
         else:
             new_kwargs[arg] = value
     return new_kwargs
+
+
+def deprecate_parameter(oldname, newname, deprecate_version, remove_version):
+    """
+    Decorator to deprecate a parameter.
+
+    Parameters
+    ----------
+    oldname : str
+        The old, deprecated parameter name.
+    newname : str
+        The new parameter name.
+    deprecate_version : str
+        The PyGMT version when the old parameter starts to be deprecated.
+    remove_version : str
+        The PyGMT version when the old parameter will be fully removed.
+
+    Examples
+    --------
+
+    >>> @deprecate_parameter("sizes", "size", "v0.0.0", "v9.9.9")
+    ... @deprecate_parameter("colors", "color", "v0.0.0", "v9.9.9")
+    ... def module(size=0, **kwargs):
+    ...     "A module that prints the arguments it received"
+    ...     print(f"size={size}, color={kwargs['color']}")
+    >>> # new names are supported
+    >>> module(size=5.0, color="red")
+    size=5.0, color=red
+    >>> # old names are supported
+    >>> module(sizes=5.0, colors="red")
+    size=5.0, color=red
+    """
+
+    def deprecator(module_func):
+        """
+        The decorator that creates our new function to work with both old and
+        new parameters.
+        """
+
+        @functools.wraps(module_func)
+        def new_module(*args, **kwargs):
+            """
+            New module instance that converts old parameters to new parameters.
+            """
+            if oldname in kwargs:
+                if newname in kwargs:
+                    raise GMTInvalidInput(
+                        f"Can't provide both '{newname}' and '{oldname}'"
+                    )
+                msg = (
+                    f"'{oldname}' is deprecated since {deprecate_version} and will"
+                    f" be removed in {remove_version}; use '{newname}' instead."
+                )
+                warnings.simplefilter("always", DeprecationWarning)  # turn off filter
+                warnings.warn(msg, DeprecationWarning, 2)
+                warnings.simplefilter("default", DeprecationWarning)  # reset filter
```

Comment:
One more thing, I don't quite get what the `warnings.simplefilter` is here for? I see it's copied from the stackoverflow answer, but I'm still wondering why we can't just do `warnings.warn`.

Reply:
I haven't read the full documentation of `warnings`, so I can't answer the question now.

Reply:
> One more thing, I don't quite get what the `warnings.simplefilter` is here for?

The `DeprecationWarning` is ignored by default. So, if you remove the `warnings.simplefilter` calls and run the example below, you won't see the warnings. That's why `warnings.simplefilter` are necessary here.
```
import pygmt
fig = pygmt.Figure()
fig.basemap(region=[-5, 5, -5, 5], projection="X10c/10c", frame=True)
fig.plot(x=[-2, 2], y=[0, 0], style="c", sizes=[1, 2])
fig.plot(x=[-2, 2], y=[2, 2], style="c", size=[1, 2])
fig.savefig("symbols.pdf")
```

---

Here are some related descriptions about `DeprecationWarning` and `FutureWarning`. If I understand it correctly, we should use `FutureWarning` instead, right?

https://docs.python.org/3/library/warnings.html

> **DeprecationWarning**
> 
> Base category for warnings about deprecated features when those warnings are intended for other **Python developers** (ignored by default, unless triggered by code in `__main__`).
> 
> **FutureWarning**
> 
> Base category for warnings about deprecated features when those warnings are **intended for end users** of applications that are written in Python.
> 
> Changed in version 3.7: Previously DeprecationWarning and FutureWarning were distinguished based on whether a feature was being removed entirely or changing its behaviour. They are now distinguished based on their intended audience and the way they’re handled by the default warnings filters.

https://www.python.org/dev/peps/pep-0565/#id13

> DeprecationWarning: reported by default for code that runs directly in the __main__ module (as such code is considered relatively unlikely to have a dedicated test suite), but hidden by default for code in other modules. The intended audience is Python developers that are at risk of upgrades to their dependencies (including upgrades to Python itself) breaking their software (e.g. developers using Python to script environments where someone else is in control of the timing of dependency upgrades).
> 
> FutureWarning: reported by default for all code. The intended audience is users of applications written in Python, rather than other Python developers (e.g. warning about use of a deprecated setting in a configuration file format).
>

Reply:
Yeah, looks like `FutureWarning` does look easier, but PEP 0565 states:

> For library and framework authors that want to ensure their API compatibility warnings are more reliably seen by their users, the recommendation is to use a custom warning class that derives from `DeprecationWarning` in Python 3.7+, and from `FutureWarning` in earlier versions.

So maybe we should stick to DeprecationWarning instead since we're Python 3.7+? I'm not so familiar with the history of DeprecationWarning/FutureWarning in Python to be honest, we should look at how other projects handle things e.g. `pandas`/`geopandas` perhaps.

Reply:
PEP 0565 also says:

> The standard library documentation will be updated to explicitly recommend the use of FutureWarning (rather than DeprecationWarning) for backwards compatibility warnings that are intended to be seen by users of an application.

Reply:
It seems both pandas and xarray use `FutureWarning` a lot:

- https://github.com/pandas-dev/pandas/blob/master/pandas/util/_depr_module.py
- https://github.com/pydata/xarray/blob/master/xarray/core/common.py

Reply:
Ok, let's use FutureWarning then, would be good to remove the `warnings.simplefilter` lines too :smile: